Adding useGCJAdapter  to decide GCJ adapter usage

rahulKQL opened the pull request:
- Added `useGCJClient` flag in BigtableOptions.
- Could be set using `google.bigtable.use.gcj.client` as argument.
- Added `BIGTABLE_USE_GCJ_CLIENT` to check availablity of this in env variable.
- Implemented `BigtableTableAdminGCJClient` with GCJ's table admin client.

andrey-qlogic commented on bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/BigtableTableAdminGCJClient.java:
```diff
@@ -9,117 +26,150 @@
   @Override
   public ApiFuture<Snapshot> getSnapshotAsync(GetSnapshotRequest request) {
-    return null;
+    return baseAdminClient.getSnapshotCallable().futureCall(request);
```
can baseAdminClient be a null ?

rahulKQL replied:
It should never be null, Also added `@Nonnull` just to verify that

andrey-qlogic commented on bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/BigtableSession.java:
```diff
@@ -402,12 +411,25 @@ public synchronized BigtableTableAdminClient getTableAdminClient() throws IOExce
+    if (options.useGCJClient()) {
+      if (adminGCJClient == null) {
+        //TODO(rahulkql): Decide If these clients need to be part of #close
+        com.google.cloud.bigtable.admin.v2.BigtableTableAdminClient adminClientV2 =
```
can you import com.google.cloud.bigtable.admin.v2.BigtableTableAdminClient ?

rahulKQL replied:
As there is already a BigtableTableAdminClient is being used in other methods, So had to use fully qualified names.

andrey-qlogic commented on bigtable-client-core-parent/bigtable-client-core/src/test/java/com/google/cloud/bigtable/grpc/TestBigtableTableAdminGCJClient.java:
```diff
@@ -0,0 +1,288 @@
+    baseClient = BaseBigtableTableAdminClient.create(baseSettings);
+    adminGCJClient = new BigtableTableAdminGCJClient(adminClient, baseClient);
+  }
+
```
remove ws

andrey-qlogic commented on bigtable-client-core-parent/bigtable-client-core/src/test/java/com/google/cloud/bigtable/grpc/TestBigtableTableAdminGCJClient.java:
```diff
@@ -0,0 +1,288 @@
+    assertEquals("secondSnapshotName",
+        actualResponse.get().getSnapshots(1).getName());
+  }
+
```
remove ws

andrey-qlogic commented on bigtable-client-core-parent/bigtable-client-core/src/test/java/com/google/cloud/bigtable/grpc/TestBigtableTableAdminGCJClient.java:
```diff
@@ -0,0 +1,288 @@
+      server = null;
+    }
+  }
+
```
remove ws

andrey-qlogic commented on bigtable-client-core-parent/bigtable-client-core/src/test/java/com/google/cloud/bigtable/grpc/TestBigtableTableAdminGCJClient.java:
```diff
@@ -0,0 +1,288 @@
+    Future<Operation> actualResponse = adminGCJClient.createTableFromSnapshotAsync(request);
+    assertTrue(actualResponse.get().getDone());
+  }
+
```
remove ws

rahulKQL commented on bigtable-client-core-parent/bigtable-client-core/src/test/java/com/google/cloud/bigtable/grpc/TestBigtableTableAdminGCJClient.java:
```diff
@@ -0,0 +1,288 @@
+    Table response = adminGCJClient.getTable(TABLE_ID);
+    assertEquals(1, response.getColumnFamilies().size());
+  }
+
```
@rahulKQL need to remove this.

sduskis committed:
Update README.md

rahulKQL committed:
Adding flag to decide GCJ adpater usage

- Added `useGCJClient` flag in BigtableOptions.
- This flag could be set using `google.bigtable.use.gcj.client` from arguments.
- Added `BIGTABLE_USE_GCJ_CLIENT` to check availablity of this in env variable.
- Implemented `BigtableTableAdminGCJClient` with GCJ's table admin client.

romlogic commented on bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/BigtableSession.java:
```diff
@@ -258,6 +266,9 @@ public BigtableSession(BigtableOptions opts) throws IOException {
     BigtableClientMetrics.counter(MetricLevel.Info, "sessions.active").inc();
 
     // Defer the creation of both the tableAdminClient until we need them.
+    //TODO(rahulkql): Need to identify which resources to initialized for GCJ's adapter case.
```
Need to identify which resource to initialize for GCJ's adapter case. 
or Need to identify which resources are to be initialized for GCJ's adapter case.
